### ISSUE TITLE
Periodic Jobs are emitted twice than expected

### ISSUE THREAD

**Issue by RidaLiu** [riverqueue/river#283]
Hi, I'm learning how to integrate river with current project. The use case is I have 2 scheduled tasks, task1 is running every minute, task2 is running every 2 minutes. Also they are set as unique in the queue level. 
the demo code is 
```go

func (CronJobArgs2) Kind() string { return "CronTask2" }

type CronTaskWorker2 struct {
	river.WorkerDefaults[CronJobArgs2]
	ctc    *conf.CronJob
	logger *zap.Logger
}

func NewCronTaskWorker2(ctc *conf.CronJob, logger *zap.Logger) *CronTaskWorker2 {
	return &CronTaskWorker2{ctc: ctc, logger: logger}
}

// biz logic
func (w *CronTaskWorker2) Work(ctx context.Context, job *river.Job[CronJobArgs2]) error {
	//biz logic, call repo and other biz modules
	w.logger.Info("Task 2")
	return nil
}

func (w *CronTaskWorker2) AddToRiverWorkers(workers *river.Workers) {
	w.logger.Info("Add worker2 to workers")
	if err := river.AddWorkerSafely(workers, w); err != nil {
		w.logger.Error("Add worker2 to workers", zap.String(lg.ErrorKey, err.Error()))
	}
}

// add cron exp
func (w *CronTaskWorker2) NewPeriodicJob() *river.PeriodicJob {
	schedule, err := cron.ParseStandard("*/2 * * * *")
	if err != nil {
		return nil
	}
	return river.NewPeriodicJob(schedule,
		func() (river.JobArgs, *river.InsertOpts) {
			return CronJobArgs2{}, nil
		},
		nil,
	)
}
func (CronTaskWorker2) InsertOpts() river.InsertOpts {
	return river.InsertOpts{
		UniqueOpts: river.UniqueOpts{
			ByQueue: true,
		},
	}
}

```

But when I create the river client and start, I fount that sometime the tasks are emitted twice near the schedule time point. I just query the data from river_job. 
like the screenshot below, task1 and task2 are emitted twice at the same schedule timestamp. I don't have much clue to debug it. 
![Snip20240321_19](https://github.com/riverqueue/river/assets/122007139/45fec6ce-aea2-47af-a86b-7749415f38e7)

**Comment by brandur** [comment]
Just before anything else, could you confirm you're on version >= 0.0.17? There was a bug specifically around the interaction of periodic and unique jobs that was fixed there.

**Comment by RidaLiu** [comment]
> Just before anything else, could you confirm you're on version >= 0.0.17? There was a bug specifically around the interaction of periodic and unique jobs that was fixed there.

current version is >= 0.0.17, the details are attached. 

123:	github.com/riverqueue/river v0.0.25 // indirect
124:	github.com/riverqueue/river/riverdriver v0.0.25 // indirect
125:	github.com/riverqueue/river/riverdriver/riverpgxv5 v0.0.25 // indirect
126:	github.com/riverqueue/river/rivertype v0.0.25 // indirect

**Comment by RidaLiu** [comment]
> ```go
> NewPeriodicJob
> ```

Thanks for your demo code. As I'm integrating the river with existing projects, so I didn't paste all the code. If the issue is cased by inserted function, it'll be quite clear in the database, as the same task will be emitted twice anytime when it meet the cron expression. Could you see my screenshot below? For the same running, at the beginning, everything is fine, shown in the green box. But  after several minutes, it's emitted twice, shown in the red box. 
![Snip20240323_22](https://github.com/riverqueue/river/assets/122007139/c72e6d62-1fbc-4a4b-88f3-c0caeb75d6d8)

another retry is similar, the issue just happened at the beginning of the running, captured in the red box. 
![Snip20240323_23](https://github.com/riverqueue/river/assets/122007139/00d93285-585c-458e-a803-bb7457b29911)

**Comment by brandur** [comment]
Could you try to modify the code above to see if you can make it reproduce the problem?

Basically we're at a point now where we have a test suite that seems to confirm this feature works, and the sample program I wrote above gives us a little additional certainty. I'm willing to believe that there is some edge bug somewhere that we haven't found, but it's hard for us to help any further without a way to reproduce it.

**Comment by RidaLiu** [comment]
> Could you try to modify the code above to see if you can make it reproduce the problem?
> 
> Basically we're at a point now where we have a test suite that seems to confirm this feature works, and the sample program I wrote above gives us a little additional certainty. I'm willing to believe that there is some edge bug somewhere that we haven't found, but it's hard for us to help any further without a way to reproduce it.


Hi, I just make some modification for my code, and have to change the unique option from true to false. 

```golang
UniqueOpts: river.UniqueOpts{ByQueue: false,},
```
and the code is 

```golang
package main

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/robfig/cron/v3"

	"github.com/riverqueue/river"
	"github.com/riverqueue/river/riverdriver/riverpgxv5"
)

type CronJobReproArgs struct{}

func (CronJobReproArgs) Kind() string { return "cron_repro1" }

type CronJobReproWorker struct {
	river.WorkerDefaults[CronJobReproArgs]
}

func (w *CronJobReproWorker) Work(ctx context.Context, job *river.Job[CronJobReproArgs]) error {
	fmt.Println("*****************")
	return nil
}

type CronJobReproArgs2 struct{}

func (CronJobReproArgs2) Kind() string { return "cron_repro2" }

type CronJobReproWorker2 struct {
	river.WorkerDefaults[CronJobReproArgs2]
}

func (w *CronJobReproWorker2) Work(ctx context.Context, job *river.Job[CronJobReproArgs2]) error {
	fmt.Println("*****************")
	return nil
}

func main() {
	ctx := context.Background()

	SecondParser := cron.NewParser(
		cron.Second | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.DowOptional | cron.Descriptor,
	)
	dbPool, err := pgxpool.New(ctx, "postgres://postgresuser:postgrespassword@postgreshost:15432")
	if err != nil {
		panic(err)
	}
	defer dbPool.Close()
	defer fmt.Printf("--- %+v\n\n", "CLOSING POOL")

	workers := river.NewWorkers()
	river.AddWorker(workers, &CronJobReproWorker2{})
	river.AddWorker(workers, &CronJobReproWorker{})
	schedule1, err := SecondParser.Parse("*/20 * * * * *")
	schedule2, err := SecondParser.Parse("*/30 * * * * *")
	if err != nil {
		panic(err)
	}

	riverClient, err := river.NewClient(riverpgxv5.New(dbPool), &river.Config{
		Logger: slog.New(slog.NewTextHandler(os.Stdout, nil)),
		PeriodicJobs: []*river.PeriodicJob{
			river.NewPeriodicJob(
				schedule1,
				func() (river.JobArgs, *river.InsertOpts) {
					return CronJobReproArgs{}, &river.InsertOpts{
						UniqueOpts: river.UniqueOpts{
							ByQueue: true,
						},
					}
				},
				&river.PeriodicJobOpts{RunOnStart: false},
			),
			river.NewPeriodicJob(
				schedule2,
				func() (river.JobArgs, *river.InsertOpts) {
					return CronJobReproArgs2{}, &river.InsertOpts{
						UniqueOpts: river.UniqueOpts{
							ByQueue: true,
						},
					}

				},
				&river.PeriodicJobOpts{RunOnStart: false},
			),
		},
		Queues: map[string]river.QueueConfig{
			river.QueueDefault: {MaxWorkers: 100},
		},
		Workers: workers,
	})
	if err != nil {
		panic(err)
	}

	fmt.Printf("--- %+v\n\n", "starting")

	if err := riverClient.Start(ctx); err != nil {
		panic(err)
	}

	fmt.Printf("--- %+v\n\n", "waiting")

	time.Sleep(15 * time.Minute)

	fmt.Printf("--- %+v\n\n", "stopping")

	if err := riverClient.Stop(ctx); err != nil {
		panic(err)
	}

	fmt.Printf("--- %+v\n\n", "stopped")
}

```
the result also have some issue after several minutes, as it should equal to 5 every minute. 
the sql code is 
```sql
select date_trunc('minute', scheduled_at) AS minute, COUNT(*),max(id) as max_id, min(id) as min_id FROM river_job where id > 4053 group by minute order by minute;
```

the issue part is captured in red box below

**Comment by brandur** [comment]
@RidaLiu Okay, thanks for taking a crack at it.

Alright, in broad strokes, if this is the only problem that's surfacing, I think this is overall a "won't fix". The periodic enqueuer tries to be as accurate as possible, but we're going to expect some minor drift for this sort of work as the performance of Go and Postgres isn't perfectly smooth over time, and the enqueue work takes a little time.

I opened #284 which changes the logic slightly so that next run times for periodic jobs are based off of their original run time instead of the current time of this run loop, and I expect this to marginally improve the problem you reported, but it won't solve it 100%.

FWIW, I ran your program a couple times and the first run produced perfectly uniform enqueues while the second produced only one deviated where a couple jobs drifted slightly from one minute to the next, but wouldn't have been off by > 1s, so also very close to uniform. I'm sure your mileage varies based on hardware, but just to show that even if the enqueuer won't be perfect, it's pretty dang close, and the drifts are so small that it'd be a fairly rare use that couldn't tolerate them.

Run no. 1:

```
         minute         | count | max_id | min_id
------------------------+-------+--------+--------
 2024-03-22 21:50:00-07 |     3 |    153 |    151
 2024-03-22 21:51:00-07 |     5 |    158 |    154
 2024-03-22 21:52:00-07 |     5 |    163 |    159
 2024-03-22 21:53:00-07 |     5 |    168 |    164
 2024-03-22 21:54:00-07 |     5 |    173 |    169
 2024-03-22 21:55:00-07 |     5 |    178 |    174
 2024-03-22 21:56:00-07 |     5 |    183 |    179
 2024-03-22 21:57:00-07 |     5 |    188 |    184
 2024-03-22 21:58:00-07 |     5 |    193 |    189
 2024-03-22 21:59:00-07 |     5 |    198 |    194
 2024-03-22 22:00:00-07 |     5 |    203 |    199
 2024-03-22 22:01:00-07 |     5 |    208 |    204
 2024-03-22 22:02:00-07 |     5 |    213 |    209
 2024-03-22 22:03:00-07 |     5 |    218 |    214
 2024-03-22 22:04:00-07 |     5 |    223 |    219
 2024-03-22 22:05:00-07 |     2 |    225 |    224
```

Run no. 2:

```
         minute         | count | max_id | min_id
------------------------+-------+--------+--------
 2024-03-22 22:06:00-07 |     5 |    230 |    226
 2024-03-22 22:07:00-07 |     5 |    235 |    231
 2024-03-22 22:08:00-07 |     5 |    240 |    236
 2024-03-22 22:09:00-07 |     5 |    245 |    241
 2024-03-22 22:10:00-07 |     5 |    250 |    246
 2024-03-22 22:11:00-07 |     5 |    255 |    251
 2024-03-22 22:12:00-07 |     5 |    260 |    256
 2024-03-22 22:13:00-07 |     5 |    265 |    261
 2024-03-22 22:14:00-07 |     5 |    270 |    266
 2024-03-22 22:15:00-07 |     5 |    275 |    271
 2024-03-22 22:16:00-07 |     5 |    280 |    276
 2024-03-22 22:17:00-07 |     5 |    285 |    281
 2024-03-22 22:18:00-07 |     7 |    292 |    286
 2024-03-22 22:19:00-07 |     3 |    295 |    293
 2024-03-22 22:20:00-07 |     5 |    300 |    296
```

**Comment by RidaLiu** [comment]
@brandur  thanks for your help, that's clear for me. But for my local cases, I didn't find 'drift' cases converge to a finial uniform scenario, but running more than expected. that's why I open this issue.

**Comment by brandur** [comment]
WIW, I was able to repro two "imbalanced" periods by leaving the program running for ~100 minutes. They're extremely rare though:

```
         minute         | count | max_id | min_id
------------------------+-------+--------+--------
 2024-03-23 00:44:00-07 |     4 |    304 |    301
 2024-03-23 00:45:00-07 |     5 |    309 |    305
 2024-03-23 00:46:00-07 |     5 |    314 |    310
 2024-03-23 00:47:00-07 |     5 |    319 |    315
 2024-03-23 00:48:00-07 |     5 |    324 |    320
 2024-03-23 00:49:00-07 |     5 |    329 |    325
 2024-03-23 00:50:00-07 |     5 |    334 |    330
 2024-03-23 00:51:00-07 |     5 |    339 |    335
 2024-03-23 00:52:00-07 |     5 |    344 |    340
 2024-03-23 00:53:00-07 |     5 |    349 |    345
 2024-03-23 00:54:00-07 |     5 |    354 |    350
 2024-03-23 00:55:00-07 |     5 |    359 |    355
 2024-03-23 00:56:00-07 |     5 |    364 |    360
 2024-03-23 00:57:00-07 |     5 |    369 |    365
 2024-03-23 00:58:00-07 |     5 |    374 |    370
 2024-03-23 00:59:00-07 |     5 |    379 |    375
 2024-03-23 01:00:00-07 |     5 |    384 |    380
 2024-03-23 01:01:00-07 |     5 |    389 |    385
 2024-03-23 01:02:00-07 |     5 |    394 |    390
 2024-03-23 01:03:00-07 |     5 |    399 |    395
 2024-03-23 01:04:00-07 |     5 |    404 |    400
 2024-03-23 01:05:00-07 |     5 |    409 |    405
 2024-03-23 01:06:00-07 |     5 |    414 |    410
 2024-03-23 01:07:00-07 |     5 |    419 |    415
 2024-03-23 01:08:00-07 |     5 |    424 |    420
 2024-03-23 01:09:00-07 |     5 |    429 |    425
 2024-03-23 01:10:00-07 |     5 |    434 |    430
 2024-03-23 01:11:00-07 |     5 |    439 |    435
 2024-03-23 01:12:00-07 |     5 |    444 |    440
 2024-03-23 01:13:00-07 |     5 |    449 |    445
 2024-03-23 01:14:00-07 |     5 |    454 |    450
 2024-03-23 01:15:00-07 |     5 |    459 |    455
 2024-03-23 01:16:00-07 |     5 |    464 |    460
 2024-03-23 01:17:00-07 |     5 |    469 |    465
 2024-03-23 01:18:00-07 |     5 |    474 |    470
 2024-03-23 01:19:00-07 |     5 |    479 |    475
 2024-03-23 01:20:00-07 |     5 |    484 |    480
 2024-03-23 01:21:00-07 |     5 |    489 |    485
 2024-03-23 01:22:00-07 |     5 |    494 |    490
 2024-03-23 01:23:00-07 |     5 |    499 |    495
 2024-03-23 01:24:00-07 |     5 |    504 |    500
 2024-03-23 01:25:00-07 |     5 |    509 |    505
 2024-03-23 01:26:00-07 |     5 |    514 |    510
 2024-03-23 01:27:00-07 |     5 |    519 |    515
 2024-03-23 01:28:00-07 |     5 |    524 |    520
 2024-03-23 01:29:00-07 |     5 |    529 |    525
 2024-03-23 01:30:00-07 |     5 |    534 |    530
 2024-03-23 01:31:00-07 |     5 |    539 |    535
 2024-03-23 01:32:00-07 |     5 |    544 |    540
 2024-03-23 01:33:00-07 |     5 |    549 |    545
 2024-03-23 01:34:00-07 |     5 |    554 |    550
 2024-03-23 01:35:00-07 |     5 |    559 |    555
 2024-03-23 01:36:00-07 |     5 |    564 |    560
 2024-03-23 01:37:00-07 |     5 |    569 |    565
 2024-03-23 01:38:00-07 |     5 |    574 |    570
 2024-03-23 01:39:00-07 |     5 |    579 |    575
 2024-03-23 01:40:00-07 |     5 |    584 |    580
 2024-03-23 01:41:00-07 |     5 |    589 |    585
 2024-03-23 01:42:00-07 |     5 |    594 |    590
 2024-03-23 01:43:00-07 |     5 |    599 |    595
 2024-03-23 01:44:00-07 |     5 |    604 |    600
 2024-03-23 01:45:00-07 |     5 |    609 |    605
 2024-03-23 01:46:00-07 |     5 |    614 |    610
 2024-03-23 01:47:00-07 |     5 |    619 |    615
 2024-03-23 01:48:00-07 |     7 |    626 |    620 <--
 2024-03-23 01:49:00-07 |     8 |    634 |    627 <--
 2024-03-23 01:50:00-07 |     5 |    639 |    635
 2024-03-23 01:51:00-07 |     5 |    644 |    640
 2024-03-23 01:52:00-07 |     5 |    649 |    645
 2024-03-23 01:53:00-07 |     5 |    654 |    650
 2024-03-23 01:54:00-07 |     5 |    659 |    655
 2024-03-23 01:55:00-07 |     5 |    664 |    660
 2024-03-23 01:56:00-07 |     5 |    669 |    665
 2024-03-23 01:57:00-07 |     5 |    674 |    670
 2024-03-23 01:58:00-07 |     5 |    679 |    675
 2024-03-23 01:59:00-07 |     5 |    684 |    680
 2024-03-23 02:00:00-07 |     5 |    689 |    685
 2024-03-23 02:01:00-07 |     5 |    694 |    690
 2024-03-23 02:02:00-07 |     5 |    699 |    695
 2024-03-23 02:03:00-07 |     5 |    704 |    700
 2024-03-23 02:04:00-07 |     5 |    709 |    705
 2024-03-23 02:05:00-07 |     5 |    714 |    710
 2024-03-23 02:06:00-07 |     5 |    719 |    715
 2024-03-23 02:07:00-07 |     5 |    724 |    720
 2024-03-23 02:08:00-07 |     5 |    729 |    725
 2024-03-23 02:09:00-07 |     5 |    734 |    730
 2024-03-23 02:10:00-07 |     5 |    739 |    735
 2024-03-23 02:11:00-07 |     5 |    744 |    740
 2024-03-23 02:12:00-07 |     5 |    749 |    745
 2024-03-23 02:13:00-07 |     5 |    754 |    750
 2024-03-23 02:14:00-07 |     5 |    759 |    755
 2024-03-23 02:15:00-07 |     5 |    764 |    760
 2024-03-23 02:16:00-07 |     5 |    769 |    765
 2024-03-23 02:17:00-07 |     5 |    774 |    770
 2024-03-23 02:18:00-07 |     5 |    779 |    775
 2024-03-23 02:19:00-07 |     5 |    784 |    780
 2024-03-23 02:20:00-07 |     5 |    789 |    785
 2024-03-23 02:21:00-07 |     2 |    791 |    790
```

FWIW, I tried leaving it running again with the fix in https://github.com/riverqueue/river/pull/284, and wasn't able to reproduce any at all. They're so rare though, that it is possible that I just got lucky the second time. We can be optimistic though.

```
         minute         | count | max_id | min_id
------------------------+-------+--------+--------
 2024-03-23 04:14:00-07 |     5 |    797 |    793
 2024-03-23 04:15:00-07 |     5 |    802 |    798
 2024-03-23 04:16:00-07 |     5 |    807 |    803
 2024-03-23 04:17:00-07 |     5 |    812 |    808
 2024-03-23 04:18:00-07 |     5 |    817 |    813
 2024-03-23 04:19:00-07 |     5 |    822 |    818
 2024-03-23 04:20:00-07 |     5 |    827 |    823
 2024-03-23 04:21:00-07 |     5 |    832 |    828
 2024-03-23 04:22:00-07 |     5 |    837 |    833
 2024-03-23 04:23:00-07 |     5 |    842 |    838
 2024-03-23 04:24:00-07 |     5 |    847 |    843
 2024-03-23 04:25:00-07 |     5 |    852 |    848
 2024-03-23 04:26:00-07 |     5 |    857 |    853
 2024-03-23 04:27:00-07 |     5 |    862 |    858
 2024-03-23 04:28:00-07 |     5 |    867 |    863
 2024-03-23 04:29:00-07 |     5 |    872 |    868
 2024-03-23 04:30:00-07 |     5 |    877 |    873
 2024-03-23 04:31:00-07 |     5 |    882 |    878
 2024-03-23 04:32:00-07 |     5 |    887 |    883
 2024-03-23 04:33:00-07 |     5 |    892 |    888
 2024-03-23 04:34:00-07 |     5 |    897 |    893
 2024-03-23 04:35:00-07 |     5 |    902 |    898
 2024-03-23 04:36:00-07 |     5 |    907 |    903
 2024-03-23 04:37:00-07 |     5 |    912 |    908
 2024-03-23 04:38:00-07 |     5 |    917 |    913
 2024-03-23 04:39:00-07 |     5 |    922 |    918
 2024-03-23 04:40:00-07 |     5 |    927 |    923
 2024-03-23 04:41:00-07 |     5 |    932 |    928
 2024-03-23 04:42:00-07 |     5 |    937 |    933
 2024-03-23 04:43:00-07 |     5 |    942 |    938
 2024-03-23 04:44:00-07 |     5 |    947 |    943
 2024-03-23 04:45:00-07 |     5 |    952 |    948
 2024-03-23 04:46:00-07 |     5 |    957 |    953
 2024-03-23 04:47:00-07 |     5 |    962 |    958
 2024-03-23 04:48:00-07 |     5 |    967 |    963
 2024-03-23 04:49:00-07 |     7 |    974 |    968
 2024-03-23 04:50:00-07 |     3 |    977 |    975
 2024-03-23 04:51:00-07 |     5 |    982 |    978
 2024-03-23 04:52:00-07 |     5 |    987 |    983
 2024-03-23 04:53:00-07 |     5 |    992 |    988
 2024-03-23 04:54:00-07 |     5 |    997 |    993
 2024-03-23 04:55:00-07 |     5 |   1002 |    998
 2024-03-23 04:56:00-07 |     5 |   1007 |   1003
 2024-03-23 04:57:00-07 |     5 |   1012 |   1008
 2024-03-23 04:58:00-07 |     5 |   1017 |   1013
 2024-03-23 04:59:00-07 |     5 |   1022 |   1018
 2024-03-23 05:00:00-07 |     5 |   1027 |   1023
 2024-03-23 05:01:00-07 |     5 |   1032 |   1028
 2024-03-23 05:02:00-07 |     5 |   1037 |   1033
 2024-03-23 05:03:00-07 |     5 |   1042 |   1038
 2024-03-23 05:04:00-07 |     5 |   1047 |   1043
 2024-03-23 05:05:00-07 |     5 |   1052 |   1048
 2024-03-23 05:06:00-07 |     5 |   1057 |   1053
 2024-03-23 05:07:00-07 |     5 |   1062 |   1058
 2024-03-23 05:08:00-07 |     5 |   1067 |   1063
 2024-03-23 05:09:00-07 |     5 |   1072 |   1068
 2024-03-23 05:10:00-07 |     5 |   1077 |   1073
 2024-03-23 05:11:00-07 |     5 |   1082 |   1078
 2024-03-23 05:12:00-07 |     5 |   1087 |   1083
 2024-03-23 05:13:00-07 |     5 |   1092 |   1088
 2024-03-23 05:14:00-07 |     5 |   1097 |   1093
 2024-03-23 05:15:00-07 |     5 |   1102 |   1098
 2024-03-23 05:16:00-07 |     5 |   1107 |   1103
 2024-03-23 05:17:00-07 |     5 |   1112 |   1108
 2024-03-23 05:18:00-07 |     5 |   1117 |   1113
 2024-03-23 05:19:00-07 |     5 |   1122 |   1118
 2024-03-23 05:20:00-07 |     5 |   1127 |   1123
 2024-03-23 05:21:00-07 |     5 |   1132 |   1128
 2024-03-23 05:22:00-07 |     5 |   1137 |   1133
 2024-03-23 05:23:00-07 |     5 |   1142 |   1138
 2024-03-23 05:24:00-07 |     5 |   1147 |   1143
 2024-03-23 05:25:00-07 |     5 |   1152 |   1148
 2024-03-23 05:26:00-07 |     5 |   1157 |   1153
 2024-03-23 05:27:00-07 |     5 |   1162 |   1158
 2024-03-23 05:28:00-07 |     5 |   1167 |   1163
 2024-03-23 05:29:00-07 |     5 |   1172 |   1168
 2024-03-23 05:30:00-07 |     5 |   1177 |   1173
 2024-03-23 05:31:00-07 |     5 |   1182 |   1178
 2024-03-23 05:32:00-07 |     5 |   1187 |   1183
 2024-03-23 05:33:00-07 |     5 |   1192 |   1188
 2024-03-23 05:34:00-07 |     5 |   1197 |   1193
 2024-03-23 05:35:00-07 |     5 |   1202 |   1198
 2024-03-23 05:36:00-07 |     7 |   1209 |   1203
 2024-03-23 05:37:00-07 |     3 |   1212 |   1210
 2024-03-23 05:38:00-07 |     7 |   1219 |   1213
 2024-03-23 05:39:00-07 |     3 |   1222 |   1220
 2024-03-23 05:40:00-07 |     5 |   1227 |   1223
 2024-03-23 05:41:00-07 |     5 |   1232 |   1228
 2024-03-23 05:42:00-07 |     5 |   1237 |   1233
 2024-03-23 05:43:00-07 |     5 |   1242 |   1238
 2024-03-23 05:44:00-07 |     5 |   1247 |   1243
 2024-03-23 05:45:00-07 |     5 |   1252 |   1248
 2024-03-23 05:46:00-07 |     5 |   1257 |   1253
 2024-03-23 05:47:00-07 |     5 |   1262 |   1258
 2024-03-23 05:48:00-07 |     5 |   1267 |   1263
 2024-03-23 05:49:00-07 |     5 |   1272 |   1268
 2024-03-23 05:50:00-07 |     5 |   1277 |   1273
 2024-03-23 05:51:00-07 |     5 |   1282 |   1278
 2024-03-23 05:52:00-07 |     5 |   1287 |   1283
```

> Just further question, for the same program, when we close it and re-run it, should we clear some data in the database?

I think it helps in these repros just to keep the database in a more easily comparable state. That's what these lines from my original sample were doing:

```
	if err := riverinternaltest.TruncateRiverTables(ctx, dbPool); err != nil {
		panic(err)
	}
```

**Comment by brandur** [comment]
@RidaLiu I just merged #284, so if you want to try River `@master`, there's a chance that it may improve the overly abundant job behavior that you observed.

**Comment by liurida** [comment]
> @RidaLiu I just merged #284, so if you want to try River `@master`, there's a chance that it may improve the overly abundant job behavior that you observed.

Thanks for your help, @brandur , I'll update my local version.
